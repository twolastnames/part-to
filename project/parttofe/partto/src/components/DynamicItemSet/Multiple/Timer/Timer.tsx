import React, { useEffect, useState } from "react";

import { DateTime, Duration, getDateTime } from "../../../../api/helpers";
import { Progress } from "../Progress/Progress";

export interface TimerProps {
  start: DateTime;
  duration: Duration;
  label: string;
  title: string;
  onClick: () => void;
}

export function Timer({ start, duration, label, title, onClick }: TimerProps) {
  const [on, setOn] = useState<number>(0);
  useEffect(() => {
    const id = setInterval(() => {
      const now = getDateTime();
      setOn(now.subtract(start).toMilliseconds());
    }, 100);
    return () => {
      clearInterval(id);
    };
  }, [start]);
  return (
    <span data-testid="Timer">
      <Progress
        onClick={onClick}
        title={title}
        label={label}
        on={on}
        total={duration.toMilliseconds()}
      />
    </span>
  );
}
