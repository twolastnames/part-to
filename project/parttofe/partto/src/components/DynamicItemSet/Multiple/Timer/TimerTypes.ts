import { DateTime } from "../../../../shared/dateTime";
import { Duration } from "../../../../shared/duration";

export interface TimerProps {
  start: DateTime;
  duration?: Duration;
  label: string;
  title: string;
  onClick: () => void;
}
