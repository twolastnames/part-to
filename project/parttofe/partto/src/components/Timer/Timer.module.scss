@import "../../App.scss";

.label {
  font: var(--font-strong-xxxlarge);
}

.clickable {
  cursor: pointer;
}
