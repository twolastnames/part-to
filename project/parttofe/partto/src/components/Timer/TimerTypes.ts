import { DateTime } from "../../shared/dateTime";
import { Duration } from "../../shared/duration";

export interface TimerProps {
  start?: DateTime;
  duration: Duration;
}
