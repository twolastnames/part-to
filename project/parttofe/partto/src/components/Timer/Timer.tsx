import React, { ReactNode, useEffect, useRef, useState } from "react";

import classes from "./Timer.module.scss";
import { TimerProps } from "./TimerTypes";
import { Ring } from "./Ring/Ring";
import { DateTime, getDateTime } from "../../shared/dateTime";
import { Duration, DurationFormat, getDuration } from "../../shared/duration";
import { Flasher } from "../Flasher/Flasher";
import { AdjustableDigit } from "./AdjustableDigit/AdjustableDigit";

const adders = [1000, 10000, 60000, 600000, 60 * 60 * 1000];

const getEnforcedDurationLabel = (
  duration: string,
  addOffset: (offset: Duration) => void,
) => (
  <>
    {
      duration
        .split("")
        .reverse()
        .reduce(
          (
            { result, adder }: { result: Array<ReactNode>; adder: number },
            character: string,
          ) => {
            const parsed = parseInt(character);
            return {
              adder: isNaN(parsed) ? adder : adder + 1,
              result: [
                isNaN(parsed) ? (
                  <span>{character}</span>
                ) : (
                  <AdjustableDigit
                    increment={
                      adders.length <= adder
                        ? undefined
                        : () => {
                            addOffset(getDuration(adders[adder]));
                          }
                    }
                    decrement={
                      adders.length <= adder
                        ? undefined
                        : () => {
                            addOffset(getDuration(-adders[adder]));
                          }
                    }
                  >
                    {character}
                  </AdjustableDigit>
                ),
                ...result,
              ],
            };
          },
          { adder: 0, result: [] },
        ).result
    }
  </>
);

const getMagnitude = (started: DateTime, duration: Duration) =>
  (getDateTime().sinceEpoch() - started.sinceEpoch()) /
  duration.toMilliseconds();

export function Timer({
  start,
  duration,
  adjustment,
  ringClasses,
}: TimerProps) {
  const [rerenderKey, setRerenderKey] = useState<number>(Math.random());
  const [started] = useState<DateTime>(start || getDateTime());
  const offset = useRef(adjustment ? adjustment.offset : getDuration(0));
  const addOffset = (value: Duration) => {
    offset.current = offset.current.add(value);
    adjustment?.setOffset(offset.current);
  };

  const getEffectiveDuration = () => duration.add(offset.current);
  const [magnitude, setMagnitude] = useState<number>(
    getMagnitude(started, getEffectiveDuration()),
  );

  useEffect(() => {
    const listener = () => {
      setMagnitude(getMagnitude(started, getEffectiveDuration()));
      setRerenderKey(Math.random());
    };
    const id = setInterval(listener, 1000);
    return () => {
      clearInterval(id);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [duration, started]);

  const labelText = started
    .add(getEffectiveDuration())
    .subtract(getDateTime())
    .format(DurationFormat.TIMER);

  const label = adjustment
    ? getEnforcedDurationLabel(labelText, addOffset)
    : labelText;
  return (
    <Ring
      key={rerenderKey}
      label={
        <span className={classes.label}>
          {magnitude > 2 ? <Flasher>{label}</Flasher> : label}
        </span>
      }
      classNames={ringClasses}
      magnitude={magnitude}
    />
  );
}
