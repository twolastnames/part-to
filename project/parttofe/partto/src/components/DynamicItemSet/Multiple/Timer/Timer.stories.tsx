import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Timer } from "./Timer";
import { ShellProvider } from "../../../../ShellProvider";

const meta: Meta<typeof Timer> = {
  component: Timer,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Timer>;

export const Simple: Story = {
  args: {},
};
