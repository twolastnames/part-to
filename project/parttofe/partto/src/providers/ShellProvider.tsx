import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
