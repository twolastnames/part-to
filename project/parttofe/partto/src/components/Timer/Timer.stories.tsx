import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Timer } from "./Timer";
import { ShellProvider } from "../../providers/ShellProvider";
import { getDateTime } from "../../shared/dateTime";
import { getDuration } from "../../shared/duration";

const meta: Meta<typeof Timer> = {
  component: Timer,
  decorators: (Story) => (
    <ShellProvider>
      <div style={{ width: "400px", height: "420px" }}>
        <Story />
      </div>
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Timer>;

export const Simple: Story = {
  args: {
    start: getDateTime(),
    duration: getDuration(65000),
  },
};

export const Hours: Story = {
  args: {
    start: getDateTime(),
    duration: getDuration(96000000),
  },
};
