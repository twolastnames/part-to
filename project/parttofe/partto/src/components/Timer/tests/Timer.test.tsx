import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../providers/ShellProvider";
import { Timer } from "../Timer";
import { getDuration } from "../../../shared/duration";

test("snapshot", () => {
  render(
    <ShellProvider>
      <Timer duration={getDuration(5000)} />
    </ShellProvider>,
  );
  const component = screen.getByTestId("Ring");
  expect(component).toMatchSnapshot();
});
