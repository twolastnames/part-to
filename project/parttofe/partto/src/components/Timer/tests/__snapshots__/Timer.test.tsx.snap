// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="ring"
  data-testid="Ring"
>
  <span
    class="show"
  >
    <div
      class="m_b32e4812 mantine-RingProgress-root"
      style="--rp-size: calc(6.25rem * var(--mantine-scale)); --rp-label-offset: calc(12.5rem * var(--mantine-scale));"
    >
      <svg
        class="m_d43b5134 mantine-RingProgress-svg"
      >
        <circle
          class="m_b1ca1fbf mantine-RingProgress-curve"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke-dasharray="0, 125.66370614359172"
          stroke-dashoffset="0"
          stroke-linecap="butt"
          stroke-width="25"
          transform="scale(1, -1) translate(0, -100)"
        />
        <circle
          class="m_b1ca1fbf mantine-RingProgress-curve"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke-dasharray="0, 125.66370614359172"
          stroke-dashoffset="125.66370614359172"
          stroke-linecap="butt"
          stroke-width="25"
          style="--curve-color: var(--secondary-color);"
        />
        <circle
          class="m_b1ca1fbf mantine-RingProgress-curve"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke-dasharray="125.66370614359172, 0"
          stroke-dashoffset="125.66370614359172"
          stroke-linecap="butt"
          stroke-width="25"
          style="--curve-color: var(--highlight-color);"
        />
      </svg>
    </div>
  </span>
  <div
    class="label"
  >
    <span
      class="label"
    >
      5
    </span>
  </div>
</div>
`;
